examples/WGAN-GP/model.py

```diff
@@ -0,0 +1,170 @@
+        dx, d_params, d_states = self.discriminator.init(rng=rng)(gx)
+
+        g_optimizer_states = self.g_optimizer.init(g_params)
+        d_optimizer_states = self.d_optimizer.init(d_params)
+
+        self.states = elegy.States(
+            g_states=g_states,
+            d_states=d_states,
+            g_params=g_params,
+            d_params=d_params,
+            g_opt_states=g_optimizer_states,
+            d_opt_states=d_optimizer_states,
+            rng=rng,
+        )
+
+    def pred_step(self, x, states):
+        z = x
+        x_fake = self.generator.apply(states.g_params, states.g_states)(z)[0]
+        return (x_fake, states)
+
+    def train_step(self, x, states):
+        # training the discriminator on every iteration
+        d_loss, d_params, d_states, d_opt_states, rng, gp = self.discriminator_step_jit(
+            x, **states
+        )
+        states = states.update(
+            d_params=d_params, d_states=d_states, d_opt_states=d_opt_states, rng=rng
+        )
+
+        self.i += 1
+        # training the generator only every 5 iterations as recommended in the original WGAN paper
+        if self.i % 5 == 0:
+            g_loss, g_params, g_states, g_opt_states, rng = self.generator_step_jit(
+                len(x), **states
+            )
+            states = states.update(
+                g_params=g_params, g_states=g_states, g_opt_states=g_opt_states, rng=rng
+            )
+        else:
+            g_loss = 0
+
+        return {"d_loss": d_loss, "g_loss": g_loss, "gp": gp}, states
+
+    def discriminator_step(
+        self, x_real, d_params, d_states, g_params, g_states, d_opt_states, rng, **_
+    ):
+        z = jax.random.normal(rng.next(), (len(x_real), 128))
+        x_fake = self.generator.apply(g_params, g_states)(z)[0]
+
+        def d_loss_fn(d_params, d_states, x_real, x_fake, rng):
+            y_real, d_params, d_states = self.discriminator.apply(d_params, d_states)(
+                x_real
+            )
+            y_fake, d_params, d_states = self.discriminator.apply(d_params, d_states)(
+                x_fake
+            )
+            y_pred = jnp.concatenate([y_real, y_fake], axis=0)
+            y_true = jnp.concatenate(
+                [jnp.ones(len(x_real)), jnp.zeros(len(x_fake))], axis=0
+            )
+            loss = -y_real.mean() + y_fake.mean()
+            gp = gradient_penalty(
+                x_real, x_fake, self.discriminator.apply(d_params, d_states), rng.next()
+            )
+            loss = loss + gp
+            return loss, (d_params, d_states, rng, gp)
+
+        (d_loss, (d_params, d_states, rng, gp)), d_grads = jax.value_and_grad(
+            d_loss_fn, has_aux=True
+        )(d_params, d_states, x_real, x_fake, rng)
+        d_grads, d_opt_states = self.d_optimizer.update(d_grads, d_opt_states, d_params)
+        d_params = optax.apply_updates(d_params, d_grads)
+
+        return d_loss, d_params, d_states, d_opt_states, rng, gp
```

Comment:
Consider accepting and returning `states` as with most `*_step` methods, the can make life easier for the caller of these methods.

Reply:
I had tried this too but got turned off by having to write `states.update(this=this, that=that, banana=banana....etc)` every time. I've now added `elegy.States.safe_update()` which is basically the opposite of `maybe_update()` and allows doing `states.safe_update(**locals())`. Less writing, happier me

Reply:
`update_known` and `update_unknown` might be better names for `maybe_update` and `safe_update`

Reply:
I like those names! You can introduce `update_known` in this PR and we can refactor to `update_unknown` in a future PR once we merge these 2 PRs that are active.

Reply:
Done.
Consider overriding operators `|` `&` `+` as shortcuts for `update_unknown`, `update_known` and `update` for even less writing: `states & locals()`